zetaclient/tss/concurrent_keysigns_tracker.go

```diff
@@ -25,26 +27,37 @@ func NewKeysignsTracker(logger zerolog.Logger) *ConcurrentKeysignsTracker {
 }
 
 // StartMsgSign is incrementing the number of active signing ceremonies as well as updating the prometheus metric
-func (k *ConcurrentKeysignsTracker) StartMsgSign() {
+//
+// Call the returned function to signify the signing is complete
+func (k *ConcurrentKeysignsTracker) StartMsgSign() func(bool) {
 	k.mu.Lock()
 	defer k.mu.Unlock()
 	k.numActiveMsgSigns++
 	metrics.NumActiveMsgSigns.Inc()
 	k.Logger.Debug().Msgf("Start TSS message sign, numActiveMsgSigns: %d", k.numActiveMsgSigns)
-}
 
-// EndMsgSign is decrementing the number of active signing ceremonies as well as updating the prometheus metric
-func (k *ConcurrentKeysignsTracker) EndMsgSign() {
-	k.mu.Lock()
-	defer k.mu.Unlock()
-	if k.numActiveMsgSigns > 0 {
-		k.numActiveMsgSigns--
-		metrics.NumActiveMsgSigns.Dec()
+	startTime := time.Now()
+
+	return func(hasError bool) {
+		k.mu.Lock()
+		defer k.mu.Unlock()
+		if k.numActiveMsgSigns > 0 {
+			k.numActiveMsgSigns--
+			metrics.NumActiveMsgSigns.Dec()
+		}
+		k.Logger.Debug().Msgf("End TSS message sign, numActiveMsgSigns: %d", k.numActiveMsgSigns)
+
+		result := "success"
+		if hasError {
+			result = "error"
+		}
+		metrics.SignLatency.With(prometheus.Labels{"result": result}).Observe(time.Since(startTime).Seconds())
 	}
-	k.Logger.Debug().Msgf("End TSS message sign, numActiveMsgSigns: %d", k.numActiveMsgSigns)
 }
 
 // GetNumActiveMessageSigns gets the current number of active signing ceremonies
 func (k *ConcurrentKeysignsTracker) GetNumActiveMessageSigns() int64 {
+	k.mu.Lock()
```

Comment:
IMO it's okay to read this metric w/o mutex

Reply:
If the variable is read while the variable is also being written the results are undefined. This variable is type `int64` (not a metric). You need a lock on both sides afaik.

In this case it probably should just be a `atomic.Int64`